src/expr/src/relation/mod.rs

```diff
@@ -343,7 +345,37 @@ impl MirRelationExpr {
                 let typ = RelationType::new(input_typ.column_types);
                 typ
             }
-            MirRelationExpr::Filter { input, .. } => input.typ(),
+            MirRelationExpr::Filter { input, predicates } => {
+                // A filter inherits the keys of its input unless the filters
+                // have reduced the input to a single row, in which case the
+                // keys of the input are `()`.
+                let input_typ = input.typ();
+                let cols_equal_to_literal = predicates
+                    .iter()
+                    .filter_map(|p| {
+                        if let MirScalarExpr::CallBinary {
+                            func: crate::BinaryFunc::Eq,
+                            expr1,
+                            expr2,
+                        } = p
+                        {
+                            if let MirScalarExpr::Column(c) = &**expr1 {
+                                if expr2.is_literal_ok() {
+                                    return Some(c);
+                                }
+                            }
+                        }
+                        None
+                    })
+                    .collect::<Vec<_>>();
+                for key_set in &input_typ.keys {
+                    if key_set.iter().all(|k| cols_equal_to_literal.contains(&k)) {
+                        let typ = RelationType::new(input_typ.column_types);
+                        return typ.with_key(vec![]);
+                    }
+                }
```

Comment:
It looks like the could be instead `key_set` minus any columns that are bound to literals, one special case of which is that they are all bound to literals. Is that wrong?

Reply:
Good point! Will modify PR accordingly.